### PR TITLE
deduce type for mem.set and mem.secureZero

### PR DESCRIPTION
**Pull request by marler8997**
Deduce type for `mem.set` and `mem.secureZero`
```zig
// current
mem.set(u8, buf, 0);
mem.secureZero(u8, buf);

// proposed
mem.set(buf, 0);
mem.secureZero(buf);
```

I've opted for a breaking change rather than creating a new method.

### CODE REVIEW

**Comment by hryx** [comment]
Nice, but this assumes that the destination slice has non-zero length.

**Comment by marler8997** [comment]
Maybe I don't understand something, but it should still work even if the destination slice has zero length.

**Comment by hryx** [comment]
Ah, my mistake -- I didn't realize the behavior was different for comptime. A small example with your proposal builds for me locally.

```zig
const std = @import("std");

pub fn main() void {
    var x: [16]u8 = undefined;
    set(x[0..], '$');
    std.debug.warn("{}\n", x[0..]);
}

fn set(dest: var, value: var) void {
    setGeneric(@typeOf(dest[0]), dest, value);
}

fn setGeneric(comptime T: type, dest: []T, value: T) void {
    for (dest) |*d|
        d.* = value;
}
```

**Comment by daurnimator** [comment]
> Maybe I don't understand something, but it should still work even if the destination slice has zero length

This was surprising to me. are we sure that's not just a bug?
Although `@typeInfo(@typeOf(dest)).Pointer.child` is more verbose, I think it may capture intention better?

> // old method still accessible with

Why keep the old methods around?

**Comment by marler8997** [comment]
> This was surprising to me. are we sure that's not just a bug?
> Although @typeInfo(@typeOf(dest)).Pointer.child is more verbose, I think it may capture intention better?

Wasn't aware of the typeinfo mechanism.  That's fine with me.

And I suppose there's not reason to keep the old methods around. I've removed them.

File: std/mem.zig [hunk 1]
```diff
@@ -248,25 +248,25 @@ pub fn copyBackwards(comptime T: type, dest: []T, source: []const T) void {
     }
 }
 
-pub fn set(comptime T: type, dest: []T, value: T) void {
+pub fn set(dest: var, value: @typeOf(dest[0])) void {
```
Comment (daurnimator) [inline std/mem.zig]: Here as well.

Comment (marler8997) [reply std/mem.zig]: Looking at the docs, should this be `Pointer.child` or `Array.child`?

Comment (daurnimator) [reply std/mem.zig]: `dest` is a slice (a type of pointer), not an array.

**Comment by marler8997** [comment]
Hey @andrewrk 

Since this one hasn't been approved yet, I'm guessing we're not sure if it's something we want.  Or maybe there is another solution to this problem in the works?  Is that the case?  If so, could you provide links and/or your thoughts or just keywords I can use to search through issues and pull requests to read up on the current state of changes like this?

After using zig a bit more, I see that this change could be applied in alot of other places as well, so I'm guessing we are taking some time to evaluate what we want to do here?

Don't feel like you need to take a bunch of time to write your thoughts here if you have already done so in other threads, just let me know what I should "search for" if that's the case.

**Comment by emekoi** [comment]
andrew is currently working on #2602 so it might be a while until he is able to review all the open pull requests.

**Comment by andrewrk** [comment]
Hey @marler8997, @emekoi is correct. I had a brief look and concluded it's not an "easy" PR as I want to give it more due consideration. Thanks for being patient

**Comment by marler8997** [comment]
No problem. Keep working on the important stuff.

**Comment by andrewrk** [comment]
Thanks for the patch, @marler8997.

I think there are some reasons to not do this:

 * previously, there would be an implicit cast when calling the functions into a slice, which is nice for array literals for example. I believe now there is a bit more bloat when calling these functions with array literals, for example. This should be observable in the test binary for behavior tests, since there are quite a few calls to `eql` with string literals.
 * we have some precedent for moving *away* from inferred types in this way.  See for example #2119.

So this is a motion to close the PR, but I'm open to a counter-argument.

**Comment by marler8997** [comment]
I assume you are talking about other functions when mentioning array literals correct? You couldn't zero array literals since they are readonly right?

I tried to gather reasons for not inferring types from #2119 but was unable. Could you enumerate them? I can imagine scenarios where you may want to provide an explicit type but it would be rare. Assuimg that's the case, I would probably opt for an explicit type function that is wrapped by an inferred type function. I think the implicit one would handle 90% of cases.

set (infers type)
setExplicit (requires explicit type)

What are you're thoughts on that?

**Comment by andrewrk** [comment]
> I assume you are talking about other functions when mentioning array literals correct? You couldn't zero array literals since they are readonly right?

I'm not sure what you mean about "zero", but maybe this helps explain:

```zig
export fn entry() void {
    _ = computeSum("abcd");
    _ = computeSum("lmnopqrs");
}

fn computeSum(x: var) usize {
    var sum: usize = 0;
    for (x) |item| sum += item;
    return sum;
}
```

```llvm
; Function Attrs: nobuiltin nounwind
define void @entry() #3 !dbg !37 {
Entry:
  %0 = call fastcc i64 @computeSum([4 x i8]* @0), !dbg !40
  %1 = call fastcc i64 @computeSum.1([8 x i8]* @1), !dbg !42
  ret void, !dbg !43
}

; Function Attrs: nobuiltin nounwind
define internal fastcc i64 @computeSum([4 x i8]* nonnull readonly align 1) unnamed_addr #3 !dbg !44 {
Entry:
  %sum = alloca i64, align 8
  %i = alloca i64, align 8
  %1 = alloca i64, align 8
  call void @llvm.dbg.declare(metadata [4 x i8]* %0, metadata !52, metadata !DIExpression()), !dbg !57
  store i64 0, i64* %sum, align 8, !dbg !58
  call void @llvm.dbg.declare(metadata i64* %sum, metadata !53, metadata !DIExpression()), !dbg !58
  store i64 0, i64* %i, align 8, !dbg !59
  call void @llvm.dbg.declare(metadata i64* %i, metadata !55, metadata !DIExpression()), !dbg !59
  br label %ForCond, !dbg !59

ForCond:                                          ; preds = %OverflowOk, %Entry
  %2 = load i64, i64* %i, align 8, !dbg !59
  %3 = icmp ult i64 %2, 4, !dbg !59
  br i1 %3, label %ForBody, label %ForEnd, !dbg !59

ForBody:                                          ; preds = %ForCond
  %4 = getelementptr inbounds [4 x i8], [4 x i8]* %0, i64 0, i64 %2, !dbg !59
  call void @llvm.dbg.declare(metadata i8* %4, metadata !56, metadata !DIExpression()), !dbg !60
  %5 = load i64, i64* %sum, align 8, !dbg !61
  %6 = load i8, i8* %4, align 1, !dbg !62
  %7 = zext i8 %6 to i64, !dbg !62
  %8 = call { i64, i1 } @llvm.uadd.with.overflow.i64(i64 %5, i64 %7), !dbg !63
  %9 = extractvalue { i64, i1 } %8, 0, !dbg !63
  %10 = extractvalue { i64, i1 } %8, 1, !dbg !63
  br i1 %10, label %OverflowFail, label %OverflowOk, !dbg !63

ForEnd:                                           ; preds = %ForCond
  %11 = load i64, i64* %sum, align 8, !dbg !64
  store i64 %11, i64* %1, !dbg !65
  ret i64 %11, !dbg !65

OverflowFail:                                     ; preds = %ForBody
  tail call fastcc void @panic(%"[]u8"* @3, %builtin.StackTrace* null), !dbg !63
  unreachable, !dbg !63

OverflowOk:                                       ; preds = %ForBody
  store i64 %9, i64* %sum, align 8, !dbg !63
  %12 = add nuw i64 %2, 1, !dbg !59
  store i64 %12, i64* %i, align 8, !dbg !59
  br label %ForCond, !dbg !59
}

; Function Attrs: nobuiltin nounwind
define internal fastcc i64 @computeSum.1([8 x i8]* nonnull readonly align 1) unnamed_addr #3 !dbg !66 {
Entry:
  %sum = alloca i64, align 8
  %i = alloca i64, align 8
  %1 = alloca i64, align 8
  call void @llvm.dbg.declare(metadata [8 x i8]* %0, metadata !74, metadata !DIExpression()), !dbg !79
  store i64 0, i64* %sum, align 8, !dbg !80
  call void @llvm.dbg.declare(metadata i64* %sum, metadata !75, metadata !DIExpression()), !dbg !80
  store i64 0, i64* %i, align 8, !dbg !81
  call void @llvm.dbg.declare(metadata i64* %i, metadata !77, metadata !DIExpression()), !dbg !81
  br label %ForCond, !dbg !81

ForCond:                                          ; preds = %OverflowOk, %Entry
  %2 = load i64, i64* %i, align 8, !dbg !81
  %3 = icmp ult i64 %2, 8, !dbg !81
  br i1 %3, label %ForBody, label %ForEnd, !dbg !81

ForBody:                                          ; preds = %ForCond
  %4 = getelementptr inbounds [8 x i8], [8 x i8]* %0, i64 0, i64 %2, !dbg !81
  call void @llvm.dbg.declare(metadata i8* %4, metadata !78, metadata !DIExpression()), !dbg !82
  %5 = load i64, i64* %sum, align 8, !dbg !83
  %6 = load i8, i8* %4, align 1, !dbg !84
  %7 = zext i8 %6 to i64, !dbg !84
  %8 = call { i64, i1 } @llvm.uadd.with.overflow.i64(i64 %5, i64 %7), !dbg !85
  %9 = extractvalue { i64, i1 } %8, 0, !dbg !85
  %10 = extractvalue { i64, i1 } %8, 1, !dbg !85
  br i1 %10, label %OverflowFail, label %OverflowOk, !dbg !85

ForEnd:                                           ; preds = %ForCond
  %11 = load i64, i64* %sum, align 8, !dbg !86
  store i64 %11, i64* %1, !dbg !87
  ret i64 %11, !dbg !87

OverflowFail:                                     ; preds = %ForBody
  tail call fastcc void @panic(%"[]u8"* @3, %builtin.StackTrace* null), !dbg !85
  unreachable, !dbg !85

OverflowOk:                                       ; preds = %ForBody
  store i64 %9, i64* %sum, align 8, !dbg !85
  %12 = add nuw i64 %2, 1, !dbg !81
  store i64 %12, i64* %i, align 8, !dbg !81
  br label %ForCond, !dbg !81
}
```

Compare that to:

```zig
export fn entry() void {
    _ = computeSum(u8, "abcd");
    _ = computeSum(u8, "lmnopqrs");
}

fn computeSum(comptime T: type, x: []const T) usize {
    var sum: usize = 0;
    for (x) |item| sum += item;
    return sum;
}
```

```llvm
; Function Attrs: nobuiltin nounwind
define void @entry() #3 !dbg !37 {
Entry:
  %0 = call fastcc i64 @computeSum(%"[]u8"* @1), !dbg !40
  %1 = call fastcc i64 @computeSum(%"[]u8"* @3), !dbg !42
  ret void, !dbg !43
}

; Function Attrs: nobuiltin nounwind
define internal fastcc i64 @computeSum(%"[]u8"* nonnull readonly align 8) unnamed_addr #3 !dbg !44 {
Entry:
  %sum = alloca i64, align 8
  %i = alloca i64, align 8
  %1 = alloca i64, align 8
  call void @llvm.dbg.declare(metadata %"[]u8"* %0, metadata !48, metadata !DIExpression()), !dbg !53
  store i64 0, i64* %sum, align 8, !dbg !54
  call void @llvm.dbg.declare(metadata i64* %sum, metadata !49, metadata !DIExpression()), !dbg !54
  store i64 0, i64* %i, align 8, !dbg !55
  call void @llvm.dbg.declare(metadata i64* %i, metadata !51, metadata !DIExpression()), !dbg !55
  %2 = getelementptr inbounds %"[]u8", %"[]u8"* %0, i32 0, i32 1, !dbg !55
  %3 = load i64, i64* %2, align 8, !dbg !55
  br label %ForCond, !dbg !55

ForCond:                                          ; preds = %OverflowOk, %Entry
  %4 = load i64, i64* %i, align 8, !dbg !55
  %5 = icmp ult i64 %4, %3, !dbg !55
  br i1 %5, label %ForBody, label %ForEnd, !dbg !55

ForBody:                                          ; preds = %ForCond
  %6 = getelementptr inbounds %"[]u8", %"[]u8"* %0, i32 0, i32 0, !dbg !55
  %7 = load i8*, i8** %6, align 8, !dbg !55
  %8 = getelementptr inbounds i8, i8* %7, i64 %4, !dbg !55
  call void @llvm.dbg.declare(metadata i8* %8, metadata !52, metadata !DIExpression()), !dbg !56
  %9 = load i64, i64* %sum, align 8, !dbg !57
  %10 = load i8, i8* %8, align 1, !dbg !58
  %11 = zext i8 %10 to i64, !dbg !58
  %12 = call { i64, i1 } @llvm.uadd.with.overflow.i64(i64 %9, i64 %11), !dbg !59
  %13 = extractvalue { i64, i1 } %12, 0, !dbg !59
  %14 = extractvalue { i64, i1 } %12, 1, !dbg !59
  br i1 %14, label %OverflowFail, label %OverflowOk, !dbg !59

ForEnd:                                           ; preds = %ForCond
  %15 = load i64, i64* %sum, align 8, !dbg !60
  store i64 %15, i64* %1, !dbg !61
  ret i64 %15, !dbg !61

OverflowFail:                                     ; preds = %ForBody
  tail call fastcc void @panic(%"[]u8"* @5, %builtin.StackTrace* null), !dbg !59
  unreachable, !dbg !59

OverflowOk:                                       ; preds = %ForBody
  store i64 %13, i64* %sum, align 8, !dbg !59
  %16 = add nuw i64 %4, 1, !dbg !55
  store i64 %16, i64* %i, align 8, !dbg !55
  br label %ForCond, !dbg !55
}
```

The second one is half the size. So that's what I meant about "bloat".

As for the second bullet point, I'll get back to you on that.

**Comment by marler8997** [comment]
Interesting.   Based on this example, it sounds you would approve of the variant I suggested?  So your example would become this:

```zig
export fn entry() void {
    _ = computeSum("abcd");
    _ = computeSum("lmnopqrs");
}

// Note that this is just an "inline" wrapper used to infer the type
inline fn computeSum(x: var) usize {
    return computeSumExplicit(x);
}

fn computeSumExplicit(comptime T: type, x: []const T) usize {
    var sum: usize = 0;
    for (x) |item| sum += item;
    return sum;
}
```

So `secureZero` would be `secureZeroExplicit` with an inline wrapper called `secureZero`.  Same thing for `set` of course.

**Comment by thejoshwolfe** [comment]
what compile error do you get when you pass a pointer instead of a slice? like `set(&foo, foo);`

**Comment by marler8997** [comment]
@thejoshwolfe, this is the error I get from this example:

```zig
export fn entry() void {
    _ = computeSum("abcd"[0..].ptr);
    _ = computeSum("lmnopqrs");
}

// Note that this is just an "inline" wrapper used to infer the type
inline fn computeSum(x: var) usize {
    return computeSumExplicit(@typeOf(x).Child, x);
}

fn computeSumExplicit(comptime T: type, x: []const T) usize {
    var sum: usize = 0;
    for (x) |item| sum += item;
    return sum;
}
```

```batch
C:\temp>zig build-obj test.zig
C:\temp\test.zig:8:49: error: expected type '[]const u8', found '[*]const u8'
    return computeSumExplicit(@typeOf(x).Child, x);
                                                ^
C:\temp\test.zig:2:19: note: called from here
    _ = computeSum("abcd"[0..].ptr);
                  ^
```

**Comment by andrewrk** [comment]
I'm not convinced to make this change.

**Comment by marler8997** [comment]
What's the reason for not supporting the interface I proposed?  `set` and `setExplicit`?  I don't think you've given any reasons yet.  The variation I proposed doesn't have any overhead since `setExplicit` will be exactly the same as the current `set` and the new `set` will be an inline wrapper.  And you said you would get back to me with reasons for why we are "moving away from inferred types".  Shouldn't we keep the PR open until the discussion is finished?  Or should I make an issue for discussion?

**Comment by andrewrk** [comment]
Indeed let's make this into a proposal rather than a PR. It's pretty clear how to implement it, the question is what should the API be.

I'm personally content with status quo so I'd like to see some reasoning on why this should change.

**Comment by marler8997** [comment]
Created a proposal/discussion thread here: https://github.com/ziglang/zig/issues/2904

**Comment by JesseRMeyer** [comment]
Does LLVM compile the 'bloat' down to reasonably tight ASM?  If not, that would be a real cause for concern, but in that case, the solution is to file a report for the LLVM folks instead of burdening Zig with inflexibility. 

In general, it is frustrating to have to specify information in a language where the compiler _already knows_ of whatever I'm specifying.  The problem does not appear to stem from any failure in Zig itself, so for Zig to have to carry the burden seems unfair to Zig users.

**Pull request closed without merging**